File: Leetcode-solution-python/countUnreachablePairsOfNodesInAnUndirectedGraph.py
from typing import List


class Solution:
    def dfs(self, node: int, adj: List[List[int]], visited: List[bool]) -> int:
        visited[node], count = True, 1
        for next_node in adj[node]:
            if visited[next_node]:
                continue
            count += self.dfs(next_node, adj, visited)
        return count

    def countPairs(self, n: int, edges: List[List[int]]) -> int:
        adj, visited, postfix, ans = [[] for i in range(n)], [False for i in range(n)], 0, 0
        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        for node in range(n):
            if visited[node]:
                continue
            temp = self.dfs(node, adj, visited)
            ans += (postfix * temp)
            postfix = temp + postfix

        return ans

File: Leetcode-solution-c-c++/countUnreachablePairsOfNodesInAnUndirectedGraph.cpp
#include <bits/stdc++.h>
using namespace std;
class Solution
{
private:
    long long dfs(int node, vector<vector<int>> &adj, vector<bool> &visited)
    {
        visited[node] = true;
        long long count = 1;
        for (auto next : adj[node])
        {
            if (visited[next])
            {
                continue;
            }
            count += dfs(next, adj, visited);
        }
        return count;
    }

public:
    long long countPairs(int n, vector<vector<int>> &edges)
    {
        vector<vector<int>> adj(n, vector<int>());
        vector<bool> visited(n, false);
        vector<long long> componentSizes;

        long long ans = 0;
        long long postfix = 0;

        for (auto edge : edges)
        {
            adj[edge[0]].push_back(edge[1]);
            adj[edge[1]].push_back(edge[0]);
        }

        long long temp = 0;
        for (int node = 0; node < n; node++)
        {
            if (visited[node])
            {
                continue;
            }
            temp = dfs(node, adj, visited);
            ans += (postfix * temp);
            postfix = temp + postfix;
        }
        ans += (postfix * temp);
        return ans;
    }
};